core/src/main/scala/filodb.core/downsample/DownsampleCardinalityManager.scala

```diff
@@ -0,0 +1,170 @@
+
+    if (filoConfig.hasPath("dataset-configs")) {
+      datasetConfig = filoConfig.getConfigList("dataset-configs")
+        .toArray().toList
+        .asInstanceOf[List[Config]]
+        .filter(x => x.getString("dataset") == datasetToSearch)
+        .headOption
+    }
+
+    val configToUse = datasetConfig match {
+      case Some(datasetConf) => Some(datasetConf)
+      case None =>
+        // use fallback to inline dataset config
+        logger.info(s"Didn't find required config for dataset=${rawDatasetRef.dataset} in config `dataset-configs`." +
+          s"Checking in config `inline-dataset-configs`")
+
+        if (filoConfig.hasPath("inline-dataset-configs")) {
+          filoConfig.getConfigList("inline-dataset-configs")
+            .toArray().toList
+            .asInstanceOf[List[Config]]
+            .filter(x => x.getString("dataset") == datasetToSearch).headOption
+        }
+        else {
+          None
+        }
+    }
+
+    configToUse match {
+      case Some(conf) =>
+        val minNumNodes = conf.getInt("min-num-nodes")
+        val numShards = conf.getInt("num-shards")
+        val result = numShards / minNumNodes
+        logger.info(s"Found the config to estimate the shards per node. minNumNodes=$minNumNodes numShards=$numShards" +
+          s" numShardsPerNode=$result")
+        result
+      case None =>
+        // NOTE: This is an Extremely UNLIKELY case, because the configs we rely on, are required for startup
+        // WHY 8? This is the current configuration of our production downsample cluster as of 06/02/2023
+        val defaultNumShardsPerNode = 8
+        logger.error(s"Could not find config for dataset=${rawDatasetRef.dataset} in configs " +
+          s"`dataset-configs` and `inline-dataset-configs`. Please check filodb config = ${filodbConfig.toString}" +
+          s" default-value=$defaultNumShardsPerNode")
+        defaultNumShardsPerNode
+    }
+  }
+
+  /**
+   * Determines if cardinality should be triggered at current hour of the day.
+   * For Example: With shardsPerNode = 8, the shard with number 0, should trigger cardinality when currentHour = 0,8,16
+   *
+   * NOTE: WHY are we using such criteria to trigger the cardinality count? -> This is done to stagger/distribute the
+   * calculation cost of cardinality count of different shards residing in the same node at different times of day.
+   * since cardinality count in downsample cluster is essentially a rebuild from scratch, which takes few mins. We don't
+   * want all the shards to be calculating this at the same time and using up all the resources of the node.
+   *
+   * @param currentShardNum Shard Number (0 - Num-Shards)
+   * @param shardsPerNode   Number of Shards per node
+   * @param currentHour     Current hour of the day (0-23)
+   * @return True if cardinality count is to be triggered, false otherwise
+   */
+  def shouldTriggerCardinalityCount(currentShardNum: Int, shardsPerNode: Int, currentHour: Int): Boolean = {
```

Comment:
How about a single-threaded scheduler that iterates through all shards once every flush-interval duration? That way, there's no room for edge-cases where we miss shards.

Reply:
Interesting point. Taking this discussion offline to learn more about it.